### ISSUE TITLE
Binding between Store model and cells

### ISSUE THREAD

**Issue by Keith-CY** [ckb-js/kuai#85]
As the [proposal](https://github.com/ckb-js/topics/blob/add-draft-of-kuai-proposal/README.md) said
 
> Note that a Store model could be a group of cells matching the same pattern and working as an entity, so it could be regarded as a virtual DApp. Say we have 2 DApps in School Roster Store models, each of them consists of many Student Store models. And we are going to build a Sports Meeting DApp, a new Sports Meeting Store model could be created and consist of partial Student Store from School A and B, respectively, it should work as same as a School Roster DApp.

A `store` model should have a `pattern` that is used to bind to specific cells. It's defined in the [architecture](https://github.com/ckb-js/kuai/issues/7) as `Resource binding` between `CKB Node` and `Store model`.
![image](https://user-images.githubusercontent.com/7271329/209507725-4660bb8f-6b17-4de7-888c-d3b1365955e3.png)

With the binding, a `store` model maps itself to specific cells and parses data from those cells into a structured data model as the state of a dapp.

Any updates on the chain should be delivered to the `store` model and keep its state up with on-chain data.

Please add the technical design of `resource binding` between `store` and `ckb node` in this issue, and then we can make a plan for the implementation.

**Comment by Keith-CY** [comment]
Could you take over this issue as a starting point for joining this project? @zhengjianhui @Daryl-L 
Any question could be appended in this issue.

**Comment by Daryl-L** [comment]
> Could you take over this issue as a starting point for joining this project? @zhengjianhui @Daryl-L Any question could be appended in this issue.

No problem~

**Comment by Daryl-L** [comment]
From my perspective, in the [IO Design](https://github.com/ckb-js/kuai/issues/15), `resource binding` module could subscribe to the block change from IO component, and then fetch the data on chain to publish to the `Store` model.

**Comment by Keith-CY** [comment]
> From my perspective, in the [IO Design](https://github.com/ckb-js/kuai/issues/15), `resource binding` module could subscribe to the block change from IO component, and then fetch the data on chain to publish to the `Store` model.

It should work, but more details are required, like how to fetch data on chain by the pattern defined in `Store`

**Comment by Daryl-L** [comment]
From this issue #2, each `Store` may have multiple states,  and each state is related to one cell. So does it mean that the relationship between `Store` and cell is many-to-many?

**Comment by Keith-CY** [comment]
> From this issue #2, each `Store` may have multiple states, and each state is related to one cell. So does it mean that the relationship between `Store` and cell is many-to-many?

Each store has a `pattern` to locate cells, it's a one-to-many relation, a cell is expected to be mapped by at most one `store` at the same time.

**Comment by zhengjianhui** [comment]
> > From this issue #2, each `Store` may have multiple states, and each state is related to one cell. So does it mean that the relationship between `Store` and cell is many-to-many?
> 
> Each store has a `pattern` to locate cells, it's a one-to-many relation, a cell is expected to be mapped by at most one `store` at the same time.

The matching pattern is a one-to-many relationship
But cell and store are a one-to-one relationship?

**Comment by Keith-CY** [comment]
> The matching pattern is a one-to-many relationship But cell and store are a one-to-one relationship?

`Store` and `cells` are one-to-many.

> Or the sudt example, say I have three cells, so by matching the pattern, should I have three stores or one?

It would be a `store` mapping to 3 cells, or to say, the store fetch data from 3 cells and parse 3 `data` into 3 `balance` and sum them up.


> From the design of the api, it should be that cells of the same schema are stored in a store. And from the use case, the path can get the specified cell ?

`Accessing to a specified cell` is the detail concealed by `store` abstraction. In the `sudt` case, user fetch `balance` by `sudt_store.get('balance')` and `store` should finish the following works:
1. sync all cells by a specific pattern, say, `sudt A` owned by `lock script B`
2. parse data from all cells, say, get several `balances`
3. return the sum of all `balances` as the `sudt A`'s balance of `user B`

User B won't know how many cells of him are used to store `sudt A`

**Comment by Daryl-L** [comment]
> > From this issue #2, each `Store` may have multiple states, and each state is related to one cell. So does it mean that the relationship between `Store` and cell is many-to-many?
> 
> Each store has a `pattern` to locate cells, it's a one-to-many relation, a cell is expected to be mapped by at most one `store` at the same time.

After reading the  issue #41 and the pr #51, I learnt that data in one cell could be separated into four `Storage`s to store, and each `Store` could hold multiple `Storage`s. But I didn't find the restriction for developers to map the same `Storage` to multiple `Store`s, so I have concluded that the relationship cells and the `Store`s is many-to-many.

**Comment by Daryl-L** [comment]
The resource binding manager is a bridge between the CKB Node and Store.

There are two parts in resource binding module, the `Manager` and the `TransactionListener`.

The `Manager` manages the relationship between the objects of `Store` and the onchain data. And it extends `Actor`, so that the `Manager` can update the data through the message queue which is defined in `Actor`.

As one cell could only be mapped to one instance of `Store`, the `Manager` is going use `Map` to store the relation, and use the `type script` and the `lock script` to match the `Store`.

When the instance of `Store` initiated, it will be registered to the `Manager` by calling `register` method. The caller must provider the pattern for manager to match the cell onchain, provider the uri of `Store` so that the `Manager` could send the message and provider the instance of `Storage` to deserialize the data.

Meanwhile, I design to provide the method `revoke` to revoke the relation. When the instance of `Store` destroyed, the `revoke` method should be called at the same time, once the `Store` is revoked, `Manager` will not update it anymore.

```typescript
type TypescriptHash = Hash
type LockscriptHash = Hash

interface Registry {
  uri: ActorURI,
  storage: ChainStorage,
}

export class Manager extends Actor {
  #registry: Map<TypescriptHash, Map<LockscriptHash, Registry>>

  private updateStore(txs: Transactions) {
    for (const tx of txs) {
      for (const input of tx.inputs) {
        this.removeFromInput(input)
      }

      for (const outputIndex in tx.outputs) {
        this.AddFromOutput(tx.outputs[outputIndex], tx.outputs_data[outputIndex], tx.witnesses[outputIndex])
      }
    }
  }
  private removeFromInput(input: Input) {
    this.call(uri, {
      type: 'remove_state', 
      remove: input.outpoint
    });
  }
  private AddFromOutput(output: Output, data: HexString, witness: HexString) {
    this.call(uri, {
      type: 'add_state',
      add: {
        `${outpoint}`: {
          data: storage.deserialize(data),
          witness: storage.deserialize(witness),
        },
      },
    })
  }
  private matchPattern(pattern: PatternType): { typeScript: Script; lockScript: Script } {}

  async register(pattern: PatternType, registry: Registry): Store {
    const { typeScript, lockScript } = this.matchPattern(pattern)
    const typeHash = utils.computeScriptHash(typeScript)
    const lockHash = utils.computeScriptHash(lockScript)
    this.#registry.get(typeHash)?.set(lockHash, await this.sync(pattern, store))
  }
  revoke(pattern: PatternType, registry: Registry): boolean {
    const { typeScript, lockScript } = this.matchPattern(pattern)
    const typeHash = utils.computeScriptHash(typeScript)
    const lockHash = utils.computeScriptHash(lockScript)
    this.#registry.get(typeHash)?.delete(lockHash)
  }
  async listen() {
    const listener = new TransactionListener()
    listener.on({
      next: this.updateStore,
    })
  }
}
```

Also I design the `TransactionListener` which is used to subscribe the transactions onchain. The `TransactionListener` extends to the `Listener<T>` which is defined in IO Component.

```typescript
import { Header } from '@ckb-lumos/lumos'

export class TransactionListener extends Listener<Header> {
  protected emit(): Promise<Header> | Header {}
}
```

Here is my design from my understanding according to the pr #51, the pr #66 and the pr #29, please have a review and point out mistakes if I was wrong. @Keith-CY

**Comment by yanguoyu** [comment]
> Here is my design from my understanding according to the pr #51, the pr #66 and the pr #29, please have a review and point out mistakes if I was wrong. @Keith-CY @zhengjianhui

I think the #registry's key should be a pattern? With a pattern there will create a store to handle cells that match this pattern, then we use the store to add or remove `store's state`.

**Comment by Keith-CY** [comment]
> > Here is my design from my understanding according to the pr #51, the pr #66 and the pr #29, please have a review and point out mistakes if I was wrong. @Keith-CY @zhengjianhui
> 
> I think the #registry's key should be a pattern? With a pattern there will create a store to handle cells that match this pattern, then we use the store to add or remove `store's state`.

From my understanding, it is a trade-off in mapping `cells` to `stores`.

The most outstanding metric of a cell is its `type script` and `lock script`. If `store` is also indexed by `type script` and `lock script`, it would be easy to locate stores affected by a cell.

The overhead is that some `stores` would be updated even they are not covered by the transaction.

**Comment by Keith-CY** [comment]
> The resource binding manager is a bridge between the CKB Node and Store.
> 
> There are two parts in resource binding module, the `Manager` and the `TransactionListener`.
> 
> The `Manager` manages the relationship between the objects of `Store` and the onchain data. And I designer the `Manager` as an `Actor`, so that the `Manager` can update the data through the message queue which is defined in `Actor`.
> 
> As one cell could only be mapped to one instance of `Store`, the `Manager` is going use `Map` to store the relation, and use the `type script` and the `lock script` to match the `Store`.
> 
> When the instance of `Store` initiated, it will be registered to the `Manager` by calling `register` method. The caller must provider the pattern for manager to match the cell onchain, provider the uri of `Store` so that the `Manager` could send the message and provider the instance of `Storage` to deserialize the data.
> 
> Meanwhile, I design to provide the method `revoke` to revoke the relation. When the instance of `Store` destroyed, the `revoke` method should be called at the same time, once the `Store` is revoked, `Manager` will not update it anymore.
> 
> ```ts
> type TypescriptHash = Hash
> type LockscriptHash = Hash
> 
> interface Registry {
>   uri: ActorURI,
>   storage: ChainStorage,
> }
> 
> export class Manager extends Actor {
>   #registry: Map<TypescriptHash, Map<LockscriptHash, Registry>>
> 
>   private updateStore(txs: Transactions) {
>     for (const tx of txs) {
>       for (const input of tx.inputs) {
>         this.removeFromInput(input)
>       }
> 
>       for (const outputIndex in tx.outputs) {
>         this.AddFromOutput(tx.outputs[outputIndex], tx.outputs_data[outputIndex], tx.witnesses[outputIndex])
>       }
>     }
>   }
>   private removeFromInput(input: Input) {
>     this.call(uri, {
>       type: 'remove_state', 
>       remove: input.outpoint
>     });
>   }
>   private AddFromOutput(output: Output, data: HexString, witness: HexString) {
>     this.call(uri, {
>       type: 'add_state',
>       add: {
>         `${outpoint}`: {
>           data: storage.deserialize(data),
>           witness: storage.deserialize(witness),
>         },
>       },
>     })
>   }
>   private matchPattern(pattern: PatternType): { typeScript: Script; lockScript: Script } {}
> 
>   async register(pattern: PatternType, registry: Registry): Store {
>     const { typeScript, lockScript } = this.matchPattern(pattern)
>     const typeHash = utils.computeScriptHash(typeScript)
>     const lockHash = utils.computeScriptHash(lockScript)
>     this.#registry.get(typeHash)?.set(lockHash, registry)
>   }
>   revoke(pattern: PatternType, registry: Registry): boolean {
>     const { typeScript, lockScript } = this.matchPattern(pattern)
>     const typeHash = utils.computeScriptHash(typeScript)
>     const lockHash = utils.computeScriptHash(lockScript)
>     this.#registry.get(typeHash)?.delete(lockHash)
>   }
>   async listen() {
>     const listener = new TransactionListener()
>     listener.on({
>       next: this.updateStore,
>     })
>   }
> }
> ```
> 
> Also I design the `TransactionListener` which is used to subscribe the transactions onchain. The `TransactionListener` extends to the `Listener<T>` which is defined in IO Component.
> 
> ```ts
> import { Header } from '@ckb-lumos/lumos'
> 
> export class TransactionListener extends Listener<Header> {
>   protected emit(): Promise<Header> | Header {}
> }
> ```
> 
> Here is my design from my understanding according to the pr #51, the pr #66 and the pr #29, please have a review and point out mistakes if I was wrong. @Keith-CY @zhengjianhui

In this design, the mapping is `type_script => lock_script => (store, cells)`

But not every cell/store has `type_script`, so there's a special `type_script` required as `null`

**Comment by Daryl-L** [comment]
> > The resource binding manager is a bridge between the CKB Node and Store.
> > There are two parts in resource binding module, the `Manager` and the `TransactionListener`.
> > The `Manager` manages the relationship between the objects of `Store` and the onchain data. And I designer the `Manager` as an `Actor`, so that the `Manager` can update the data through the message queue which is defined in `Actor`.
> > As one cell could only be mapped to one instance of `Store`, the `Manager` is going use `Map` to store the relation, and use the `type script` and the `lock script` to match the `Store`.
> > When the instance of `Store` initiated, it will be registered to the `Manager` by calling `register` method. The caller must provider the pattern for manager to match the cell onchain, provider the uri of `Store` so that the `Manager` could send the message and provider the instance of `Storage` to deserialize the data.
> > Meanwhile, I design to provide the method `revoke` to revoke the relation. When the instance of `Store` destroyed, the `revoke` method should be called at the same time, once the `Store` is revoked, `Manager` will not update it anymore.
> > ```ts
> > type TypescriptHash = Hash
> > type LockscriptHash = Hash
> > 
> > interface Registry {
> >   uri: ActorURI,
> >   storage: ChainStorage,
> > }
> > 
> > export class Manager extends Actor {
> >   #registry: Map<TypescriptHash, Map<LockscriptHash, Registry>>
> > 
> >   private updateStore(txs: Transactions) {
> >     for (const tx of txs) {
> >       for (const input of tx.inputs) {
> >         this.removeFromInput(input)
> >       }
> > 
> >       for (const outputIndex in tx.outputs) {
> >         this.AddFromOutput(tx.outputs[outputIndex], tx.outputs_data[outputIndex], tx.witnesses[outputIndex])
> >       }
> >     }
> >   }
> >   private removeFromInput(input: Input) {
> >     this.call(uri, {
> >       type: 'remove_state', 
> >       remove: input.outpoint
> >     });
> >   }
> >   private AddFromOutput(output: Output, data: HexString, witness: HexString) {
> >     this.call(uri, {
> >       type: 'add_state',
> >       add: {
> >         `${outpoint}`: {
> >           data: storage.deserialize(data),
> >           witness: storage.deserialize(witness),
> >         },
> >       },
> >     })
> >   }
> >   private matchPattern(pattern: PatternType): { typeScript: Script; lockScript: Script } {}
> > 
> >   async register(pattern: PatternType, registry: Registry): Store {
> >     const { typeScript, lockScript } = this.matchPattern(pattern)
> >     const typeHash = utils.computeScriptHash(typeScript)
> >     const lockHash = utils.computeScriptHash(lockScript)
> >     this.#registry.get(typeHash)?.set(lockHash, registry)
> >   }
> >   revoke(pattern: PatternType, registry: Registry): boolean {
> >     const { typeScript, lockScript } = this.matchPattern(pattern)
> >     const typeHash = utils.computeScriptHash(typeScript)
> >     const lockHash = utils.computeScriptHash(lockScript)
> >     this.#registry.get(typeHash)?.delete(lockHash)
> >   }
> >   async listen() {
> >     const listener = new TransactionListener()
> >     listener.on({
> >       next: this.updateStore,
> >     })
> >   }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Also I design the `TransactionListener` which is used to subscribe the transactions onchain. The `TransactionListener` extends to the `Listener<T>` which is defined in IO Component.
> > ```ts
> > import { Header } from '@ckb-lumos/lumos'
> > 
> > export class TransactionListener extends Listener<Header> {
> >   protected emit(): Promise<Header> | Header {}
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Here is my design from my understanding according to the pr #51, the pr #66 and the pr #29, please have a review and point out mistakes if I was wrong. @Keith-CY @zhengjianhui
> 
> In this design, the mapping is `type_script => lock_script => (store, cells)`
> 
> But not every cell/store has `type_script`, so there's a special key required as `type script is null`

That's right, for cells which don't have the `type script`, I just use `null` as the key to map cells. So that the type of `type script` should be modified to `type TypescriptHash = Hash | 'null'`.

**Comment by Daryl-L** [comment]
> > Here is my design from my understanding according to the pr #51, the pr #66 and the pr #29, please have a review and point out mistakes if I was wrong. @Keith-CY @zhengjianhui
> 
> I think the #registry's key should be a pattern? With a pattern there will create a store to handle cells that match this pattern, then we use the store to add or remove `store's state`.

From my understanding, the cells could only be mapped to store by `type script` and `lock script`, we don't need anything else to map cells. So I use `Map` with hash of `type script` and `lock script` to match cells.

And from what you said, this `Resource Binding` would initiate data of store and return it when the `Store` registered, right?

**Comment by zhengjianhui** [comment]
> ```ts
> export class Manager extends Actor {
> ```

Here I have a question, how to handle if more than one actor uses the same cell? I have to understand that the better way is to use composition and dependency injection?

**Comment by Daryl-L** [comment]
> > ```ts
> > export class Manager extends Actor {
> > ```
> 
> Here I have a question, how to handle if more than one actor uses the same cell? I think extends is used here, I understand a better way is to use composition and dependency injection?

Here, the `Manager` extends `Actor` means that `Manager` runs as an `Actor` to send "add_state" or "remove_state" message to `Store` by using message queue which is defined in `Actor`.

**Comment by Daryl-L** [comment]
Here to answer why to achieve the `Manager` extended from `Actor` and how to use this `Manager` in an application.

As the issue #4 designed, the actor system has a message queue for `Actor`s to deliver messages, the `Manager` is designed as an `Actor` so that it can send message to other `Actor`s like `Store`s etc. By using `Registry` which is achieved in pr #29, the singleton of `Manager` will be register to the `Registry`, as a part of the actor system.

The `Manager` may be used like this.

```typescript
const registry = new Registry();
registry.load();

const resourceBindingManager = registry.find('local://resource-binding');
const store = registry.find('local://sotre');
store.call('local://resource-binding', ${registerMessage});

// And then the resource binding manager will fetch the latest data from onchain when receive this register message, and send the data by message to the store.

// before destroying store, the revoke message will be send to resource binding manager to revoke the store
store.call('local://resource-binding', ${revokeMessage});
```

**Comment by Daryl-L** [comment]
The message for updating cell is defined here.

```typescript
interface Message {
  type: 'update_cell',
  update: {
    witness: HexString,
    cell: Cell,
}
```

The `Cell` is defined in lumos.

```typescript
interface Cell {
    cellOutput: {
        capacity: HexNumber;
        lock: Script;
        type?: Script;
    };
    data: HexString;
    outPoint?: OutPoint;
    blockHash?: Hash;
    blockNumber?: HexNumber;
}
```

The `HexString` is also defined in lumos.

```typescript
type HexString = string;
```

And for the `remove_state`, just use what is defined in pr #51.

Is that OK? @yanguoyu

**Comment by yanguoyu** [comment]
> The message for updating cell is defined here.
> 
> ```ts
> interface Message {
>   type: 'update_cell',
>   update: {
>     witness: HexString,
>     cell: Cell,
> }
> ```
> 
> The `Cell` is defined in lumos.
> 
> ```ts
> interface Cell {
>     cellOutput: {
>         capacity: HexNumber;
>         lock: Script;
>         type?: Script;
>     };
>     data: HexString;
>     outPoint?: OutPoint;
>     blockHash?: Hash;
>     blockNumber?: HexNumber;
> }
> ```
> 
> The `HexString` is also defined in lumos.
> 
> ```ts
> type HexString = string;
> ```
> 
> And for the `remove_state`, just use what is defined in pr #51.
> 
> Is that OK? @yanguoyu

OK, no problem.

**Comment by Keith-CY** [comment]
> > The message for updating cell is defined here.
> > ```ts
> > interface Message {
> >   type: 'update_cell',
> >   update: {
> >     witness: HexString,
> >     cell: Cell,
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The `Cell` is defined in lumos.
> > ```ts
> > interface Cell {
> >     cellOutput: {
> >         capacity: HexNumber;
> >         lock: Script;
> >         type?: Script;
> >     };
> >     data: HexString;
> >     outPoint?: OutPoint;
> >     blockHash?: Hash;
> >     blockNumber?: HexNumber;
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The `HexString` is also defined in lumos.
> > ```ts
> > type HexString = string;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And for the `remove_state`, just use what is defined in pr #51.
> > Is that OK? @yanguoyu
> 
> OK, no problem.

Will this update be appended in PR https://github.com/ckb-js/kuai/pull/51 or create another one after https://github.com/ckb-js/kuai/pull/51 is merged

**Comment by yanguoyu** [comment]
> Will this update be appended in PR #51 or create another one after #51 is merged(2 approvals already)

I will create a new PR to add this type.

**Comment by yanguoyu** [comment]
I will change the message Interface to 
```typescript
export type UpdateStorageValue = {
  witness: HexString,
  cell: Cell,
}

export type StoreMessage = {
  type: 'remove_cell',
  value: OutPointString[]
} | {
  type: 'update_cell'
  value: UpdateStorageValue
}

```
Then the Interface is unified for callers.

**Comment by Keith-CY** [comment]
PR: https://github.com/ckb-js/kuai/pull/100
